### ISSUE TITLE
BatchNorm prevents 1-dimensional arrays as input

### ISSUE THREAD

**Issue by HenriDeh** [FluxML/Flux.jl#1125]
Let us take the Chain example given in the [documentation](https://fluxml.ai/Flux.jl/stable/models/layers/#Flux.BatchNorm) on BatchNorm layers:

```
m = Chain(
  Dense(28^2, 64),
  BatchNorm(64, relu),
  Dense(64, 10),
  BatchNorm(10),
  softmax)
```

and the equivalent network without batch normalization layers:
```
m2 = Chain(Dense(28^2,64,relu), Dense(64,10),softmax)
```
`m` does not accept a 1-dimensional array but `m2` does. One would expect both chains to behave identically.

```
m(rand(28^2)) 
ERROR: arraysize: dimension out of range
m(hcat(rand(28^2))) # works ok

m2(rand(28^2)) 
m2(hcat(rand(28^2))) # both work ok
```

This was discussed in previous issues like #976 and judged as "not an issue". If not a bug, that's a missing feature in Flux in my opinion. Chains are no longer able to compute with `AbstractArrays` as input. Adding BatchNorm layers to my project required me to change code at several places. At the very least, we should change the error message.

**Comment by bhvieira** [comment]
Flux quite explicitly expects a batch dimension, even if singleton. A vector would be a vector of 1-D data, not a N-D datum. I think Flux behavior is quite correct, avoiding ambiguities.

**Comment by CarloLucibello** [comment]
@bhvieira is right, if we start making guesses about the presence or not of batch dimension we would have a lot of ambiguities
